### PR TITLE
Preprocessor

### LINKED ISSUE #63: Support custom postprocessors
Add `postprocessor` key to the project specification that will be a module implementing a `Bundlex.Project.Postprocessor` behaviour with a single callback `process`, that will be able to extend natives specifications with custom parameters

### PR DESCRIPTION
**Pull request by mat-hek**
closes #63

### CODE REVIEW

File: lib/bundlex/native.ex [hunk 1]
```diff
@@ -76,61 +94,57 @@ defmodule Bundlex.Native do
   end
 
   defp resolve_native(config, erlang, src_path, platform) do
-    native_interfaces = Keyword.get(config.config, :interface, [])
+    with {:ok, native} <- parse_native(config, src_path, nil) do
```
[the remark this answers is not in the document]
Comment (mat-hek) [reply lib/bundlex/native.ex:97]: The config in the line 130 is not the same config, it's preprocessed in 122-127. I agree that's counterintuitive though, will refactor

File: lib/bundlex/native.ex [hunk 2]
```diff
@@ -16,11 +19,12 @@ defmodule Bundlex.Native do
...
           pkg_configs: [String.t()],
           sources: [String.t()],
-          deps: [String.t()],
+          deps: [t],
           compiler_flags: [String.t()],
           linker_flags: [String.t()],
           language: :c | :cpp,
-          interface: [interface_t] | interface_t
+          interface: interface_t | nil,
+          precompilers: Precompiler.t() | [Precompiler.t()]
```
[the remark this answers is not in the document]
Comment (mat-hek) [reply lib/bundlex/native.ex:27]: In this case, it's always a list, fixed the type

[the remark this answers is not in the document]
Comment (mickel8) [reply lib/bundlex/native.ex:27]: I missed this with struct in `project.ex`. So shouldn't we change this to `precompilers: [Precompiler.t()]`?

File: lib/bundlex/project/precompiler.ex [hunk 3]
```diff
@@ -0,0 +1,20 @@
+defmodule Bundlex.Project.Precompiler do
```
Comment (bblaszkow06) [inline lib/bundlex/project/precompiler.ex:1]: Wisdom of BBB: Where does the name `Precompiler` come from? To me it seems more like `Preprocessor` or even `Processor`

Comment (mat-hek) [reply lib/bundlex/project/precompiler.ex:1]: I started from the processor, but it didn't sound well, so switched to precompiler. Agree that preprocessor would be better, but not sure if I worth changing

File: lib/bundlex/project/preprocessor.ex [hunk 4]
```diff
@@ -1,20 +1,20 @@
-defmodule Bundlex.Project.Precompiler do
+defmodule Bundlex.Project.Preprocessor do
   @moduledoc """
   Behaviour for precompiling Bundlex projects.
 
   Precompiling may involve either generating additional resources or altering the project itself.
-  Currently, precompiling native configuration (`c:precompile_native_config/3`)
-  and parsed natives (`c:precompile_native/1`) is supported.
+  Currently, precompiling native configuration (`c:preprocess_native_config/3`)
```
Comment (mickel8) [inline lib/bundlex/project/preprocessor.ex:6]: precompiling -> preprocessing

File: lib/bundlex/project/preprocessor.ex [hunk 5]
```diff
@@ -1,20 +1,20 @@
-defmodule Bundlex.Project.Precompiler do
+defmodule Bundlex.Project.Preprocessor do
   @moduledoc """
   Behaviour for precompiling Bundlex projects.
```
Comment (mickel8) [inline lib/bundlex/project/preprocessor.ex:3]: precompiling -> preprocessing

**Comment by bblaszkow06** [comment]
One more preprocessor left in PR name - make sure to change it before merging (or when merging)

**Pull request merged**